sdk/eventhub/Azure.Messaging.EventHubs/design/event-processor{T}-proposal.md

```diff
@@ -0,0 +1,520 @@
+    {
+        await DoSomethingWithEvents(events).ConfigureAwait(false);
+        Log.Information($"Received event batch for partition: { partition.PartitionId }");
+    }
+    
+    protected override Task OnProcessingErrorAsync(Exception exception, CustomPartition partition, string operationDescription, CancellationToken cancellationToken = default)
+    {
+        Log.Error(exception, $"Error occured for partition: { partition.PartitionId } during { operationDescription }.");
+        return Task.CompletedTask;
+    }
+    
+    // Storage integration
+    protected override Task<IEnumerable<EventProcessorCheckpoint>> ListCheckpointsAsync(CancellationToken cancellationToken) =>
+        CustomStorage.GetCheckpointsAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup);
+    
+    protected override Task<IEnumerable<PartitionOwnership>> ListOwnershipAsync(CancellationToken cancellationToken) =>
+        CustomStorage.GetOwnershipAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup, Identifier);
+    
+    protected override Task<IEnumerable<PartitionOwnership>> ClaimOwnershipAsync(IEnumerable<PartitionOwnership> desiredOwnership, CancellationToken cancellationToken) =>
+        CustomStorage.TryUpdateOwnershipAsync(desiredOwnership);
+}
+```
+
+### Create a processor and perform basic event processing
+
+```csharp
+public class CustomProcessor : EventProcessor<EventProcessorPartition> 
+{
+    // ... implementation
+}
+
+// Create the processor
+var processor = new CustomProcessor(
+    100, 
+    EventHubConsumerClient.DefaultConsumerGroupName,
+    "somenamespace.thing.servicebus.net",
+    "myAwesomeHub",
+    new DefaultAzureCredential());
+    
+// Processing for events and errors is defined as part
+// of the custom processor class.  There are no handlers
+// to hook up here.
+
+await processor.StartProcessingEventsAsync();
+
+// Developer code can continue to perform actions;
+// processing is taking place in the background.
+
+using var cancellationSource = new TokenCancellationSource();
+cancellationSource.CancelAfter(Timespan.FromMinutes(5));
+
+try
+{
+    await Task.Delay(Timeout.Infinite, cancellationSource.Token);
+}
+catch (TaskCanceledException)
+{
+    // Expected; no action is necessary.
+}
+
+await processor.StopProcessingAsync();
+```
+
+### Create a processor with custom options
+
+```csharp
+public class CustomProcessor : EventProcessor<EventProcessorPartition>
+{
+    // ... implementation
+}
+
+var options = new EventProcessorOptions
+{
+    Identifier = "RedFive",
+    PrefetchCount = 300,
+    TrackLastEnqueuedEventProperties = true
+};
+
+// Specify the default starting position when no checkpoint was found.
+//
+// Overriding the default starting position for specific partitions an be achieved by 
+// returning the desired location when ListCheckpointsAsync is called.
```

Comment:
You're saying that, since the users are now responsible for implementing `ListCheckpointsAsync`, they themselves can use the method to override the initial position and, for that reason, we don't need them to set up the position on initialization?

Reply:
Correct.   If they wish to set a specific default location for a specific set of partitions, they can lie about the checkpoints that exist.